overlord/devicestate/devicestate.go

```diff
@@ -433,7 +435,48 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
 		(ms.newModelSnap.SnapType == "kernel" || ms.newModelSnap.SnapType == "gadget")
 }
 
-func remodelEssentialSnapTasks(ctx context.Context, st *state.State, ms modelSnapsForRemodel, deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
+type remodelOptions struct {
+	localSnapsRequired bool
+}
+
+func (ro *remodelOptions) UpdateWithDeviceContext(st *state.State,
+	localSi *snap.SideInfo, path, snapName, channel string,
+	userID int, snapStateFlags snapstate.Flags,
+	deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
+	logger.Debugf("snap %s track changed", snapName)
+	if ro.localSnapsRequired {
+		if localSi == nil {
+			return nil, fmt.Errorf("no file provided for %q snap (track changed)", snapName)
+		}
+		return snapstateUpdatePathWithDeviceContext(st, localSi, path, snapName,
+			&snapstate.RevisionOptions{Channel: channel},
+			userID, snapStateFlags, deviceCtx, fromChange)
+	}
+	// new modes specifies the same snap, but with a new channel
+	return snapstateUpdateWithDeviceContext(st, snapName,
+		&snapstate.RevisionOptions{Channel: channel},
+		userID, snapStateFlags, deviceCtx, fromChange)
+}
+
+func (ro *remodelOptions) InstallWithDeviceContext(ctx context.Context, st *state.State,
+	localSi *snap.SideInfo, path, snapName, channel string, userID int,
+	snapStateFlags snapstate.Flags, deviceCtx snapstate.DeviceContext,
+	fromChange string) (*state.TaskSet, error) {
+	logger.Debugf("snap %s needs install", snapName)
+	if ro.localSnapsRequired {
+		if localSi == nil {
+			return nil, fmt.Errorf("no file provided for %q snap", snapName)
+		}
+		return snapstateInstallPathWithDeviceContext(st, localSi, path, snapName,
+			&snapstate.RevisionOptions{Channel: channel},
+			userID, snapStateFlags, deviceCtx, fromChange)
+	}
+	return snapstateInstallWithDeviceContext(ctx, st, snapName,
+		&snapstate.RevisionOptions{Channel: channel},
+		userID, snapStateFlags, deviceCtx, fromChange)
+}
+
+func remodelEssentialSnapTasks(ctx context.Context, st *state.State, ms modelSnapsForRemodel, deviceCtx snapstate.DeviceContext, fromChange string, remodelOpts remodelOptions, localSi *snap.SideInfo, path string) (*state.TaskSet, error) {
```

Comment:
this probably should have:

remodeOpts, deviceCtx, fromChange

as last arguments for consistency